### PR TITLE
feat(auth): 企业 LDAP 登录与身份绑定

### PR DESCRIPTION
**Pull request by jangrui**
基于 PR #283 的 LDAP 目录登录实现，按 #437 review 要求完整实现企业级 LDAP 认证。

认证与身份：
- LDAP 作为本地认证失败后的回退，支持 OpenLDAP（entryUUID）/ AD（objectGUID）稳定目录标识
- 通过 IdentityBinding 绑定本地账号，首次登录持久化、重复登录按 subject 命中，不重复建号
- 禁止邮箱静默合并：邮箱碰撞返回 409，不接管既有本地/OAuth 账号
- 新增显式绑定端点（POST /api/v1/auth/ldap/bind），解决邮箱冲突后的自助绑定入口
- 可配置属性映射（username/displayName/email/subject）与登录属性同步

安全与配置：
- LDAPS 自定义信任库（EnvironmentPostProcessor 合并 CA），支持企业自签证书
- 错误四分类（用户不存在/密码错误/目录不可用/TLS 证书）+ 中英文消息键
- LDAP 条件化装配（disabled 不初始化连接），日志脱敏 URL 凭据，用户名正则防注入
- 连接超时可配，移除 LdapTemplate 死依赖

前端：
- 设置页 LDAP 绑定卡片
- api-error 展示后端业务消息（账号禁用/目录不可用/TLS 错误）

测试：
- LdapAuthService 专属单元测试（身份/碰撞/同步/禁用）
- 真实 OpenLDAP Testcontainers 集成测试（登录全链路、LDAPS、并发首登、目录不可用、disabled 启动）

### CODE REVIEW

File: server/skillhub-auth/src/main/java/com/iflytek/skillhub/auth/ldap/LdapAuthService.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+        UserAccount user = null;
+        
+        // Try to find by email first
+        if (email != null && !email.isEmpty()) {
+            user = userAccountRepository.findByEmailIgnoreCase(email.toLowerCase()).orElse(null);
+        }
+
+        // If not found, create a new user
+        if (user == null) {
```
[the remark this answers is not in the document]
Comment (jangrui) [reply server/skillhub-auth/src/main/java/com/iflytek/skillhub/auth/ldap/LdapAuthService.java]: ✅ **已修复**

修复方案：使用 `ldap:{username}@internal` 作为唯一标识符，确保同一 LDAP 用户始终映射到同一个本地账号，防止重复账号创建和用户数据丢失。

提交: d8b507b8

File: server/skillhub-auth/src/test/java/com/iflytek/skillhub/auth/local/LocalAuthServiceTest.java [hunk 2]
```diff
@@ -238,4 +251,66 @@ void register_rejectsInvalidEmailFormat() {
...
+        UserAccount ldapUser = new UserAccount("usr_ldap", "ldapuser", "ldapuser@example.com", null);
+        ldapUser.setStatus(UserStatus.ACTIVE);
+        PlatformPrincipal ldapPrincipal = new PlatformPrincipal(
+            "usr_ldap",
+            "ldapuser",
+            "ldapuser@example.com",
+            Set.of("USER"),
+            Set.of(),
+            Set.of()
+        );
```
[the remark this answers is not in the document]
Comment (jangrui) [reply server/skillhub-auth/src/test/java/com/iflytek/skillhub/auth/local/LocalAuthServiceTest.java:289]: ✅ **已修复**

修复方案：更正 PlatformPrincipal 构造函数参数顺序为：
`(userId, displayName, email, avatarUrl, oauthProvider, roles)`

与 LocalAuthService 和 LdapAuthService 中的调用保持一致。

提交: d8b507b8

File: server/skillhub-auth/src/main/java/com/iflytek/skillhub/auth/ldap/LdapAuthService.java [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+            if (ldapProperties.getUsername() != null && !ldapProperties.getUsername().isEmpty()) {
+                env.put(Context.SECURITY_PRINCIPAL, ldapProperties.getUsername());
+                env.put(Context.SECURITY_CREDENTIALS, ldapProperties.getPassword());
+            }
+            
+            DirContext ctx = new InitialDirContext(env);
+            Attributes attrs = ctx.getAttributes(new LdapName(userDn));
+            ctx.close();
+            return attrs;
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (jangrui) [reply server/skillhub-auth/src/main/java/com/iflytek/skillhub/auth/ldap/LdapAuthService.java:500]: ✅ **已修复**

修复方案：在 finally 块中显式关闭 NamingEnumeration 和 DirContext，防止资源泄露。

```java
finally {
    if (results != null) {
        try {
            results.close();
        } catch (Exception e) {
            log.warn("Failed to close LDAP search results", e);
        }
    }
    closeContext(ctx);
}
```

提交: d8b507b8

File: server/skillhub-auth/src/main/java/com/iflytek/skillhub/auth/ldap/LdapAuthService.java [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+                           LdapTemplate ldapTemplate,
+                           UserAccountRepository userAccountRepository,
+                           UserRoleBindingRepository userRoleBindingRepository,
+                           GlobalNamespaceMembershipService globalNamespaceMembershipService) {
+        this.ldapProperties = ldapProperties;
+        this.ldapTemplate = ldapTemplate;
+        this.userAccountRepository = userAccountRepository;
+        this.userRoleBindingRepository = userRoleBindingRepository;
+        this.globalNamespaceMembershipService = globalNamespaceMembershipService;
+    }
```
[the remark this answers is not in the document]
Comment (jangrui) [reply server/skillhub-auth/src/main/java/com/iflytek/skillhub/auth/ldap/LdapAuthService.java]: ℹ️️ **不采纳**

原因：
1. **保持代码一致性**：原 PR #283 使用手动 JNDI 操作，为保持一致性，我们沿用了该模式
2. **资源管理已修复**：通过显式关闭 `NamingEnumeration` 和 `DirContext`，资源泄露问题已解决
3. **最小化变更原则**：作为 bug 修复 PR，应尽量减少架构性变更
4. **功能完整性**：手动 JNDI 操作提供了更细粒度的控制，便于调试和错误处理

未来改进建议：如果进行大规模重构，可以考虑迁移到 `LdapTemplate`。

提交: d8b507b8

**Comment by jangrui** [comment]
@dongmucat 

您好！这个 PR 修复了 PR #283 的测试问题，包括：
- 明确配置 LDAP mock 默认行为
- 添加 LDAP 回退功能测试
- 修复 Gemini Code Assist 指出的关键问题（重复账号、资源泄露）

当前有 2 个 workflows 需要您的批准：
- PR Tests
- PR E2E Tests

烦请批准运行，谢谢！

**Comment by dongmucat** [comment]
> @dongmucat
> 
> 您好！这个 PR 修复了 PR #283 的测试问题，包括：
> 
> * 明确配置 LDAP mock 默认行为
> * 添加 LDAP 回退功能测试
> * 修复 Gemini Code Assist 指出的关键问题（重复账号、资源泄露）
> 
> 当前有 2 个 workflows 需要您的批准：
> 
> * PR Tests
> * PR E2E Tests
> 
> 烦请批准运行，谢谢！

好的，已批准

**Comment by XiaoSeS** [comment]
@jangrui 辛苦了，接力推进 LDAP 这个 PR 不容易，几个关键修复都很到位 👍 我从外部看了一下 #283 和 #437 的对比，整理一些想法供参考。

**相比 #283 已经修好的几处都很关键：**

- `NamingEnumeration` / `DirContext` 在 `finally` 里关闭 — 高并发场景下的资源泄露隐患解决了
- 用 `ldap:{username}@internal` 作 email key 避免重复账号 — 这个思路挺巧
- 测试里 `PlatformPrincipal` 构造器顺序修正 + `ldapProperties.isEnabled()` mock 显式化，CI 失败的根因都覆盖到了
- 补的 2 个 LDAP fallback 测试用例守住了核心新逻辑

**一些可以再讨论的点：**

1. **`spring-boot-starter-data-ldap` 的 AutoConfiguration 副作用**

   这个依赖会在启动时尝试初始化 LDAP 连接，即使 `SKILLHUB_LDAP_ENABLED=false`。`application.yml` 里加的 `management.health.ldap.enabled: false` 只关掉了健康检查端点，AutoConfiguration 本身仍会触发。对于不用 LDAP 的存量部署可能产生启动日志噪音或连接错误。

   一个相对小的改法是在 `application.yml` 里补一行：

   ```yaml
   spring:
     ldap:
       urls: ""
   ```

   或者把 `LdapTemplate` 的 Bean 改成按 `skillhub.ldap.enabled` 条件化创建。完全理解你"最小化变更"的考量，但这个不属于架构重构，更接近修一个隐性 bug。

2. **`ldap:{username}@internal` 这个标识符约定**

   解决重复账号的方向是对的，不过这个格式现在是隐式约定，将来读到 `email = ldap:xxx@internal` 的人会很困惑。能否在 `LdapAuthService` 里加几行注释说明这是 LDAP 用户的内部唯一键？另外如果 LDAP 里两个不同 OU 下有同 `uid` 的用户，会撞 key —— 是否考虑用 `entryUUID` / `objectGUID` 之类的稳定标识？这个不阻塞合并，作为后续改进点提一下。

3. **debug 日志可能泄露 LDAP URL 里的密码**

   `LdapAuthService.login()` 和 `LocalAuthService.login()` 里都有 `log.debug("LDAP URL: {}", ldapProperties.getUrl())`。如果运维把 URL 写成 `ldap://binduser:password@host` 这种形式（不常见但合法），debug 模式下会明文打印。建议要么去掉这行，要么只打印 host 部分。

4. **PR 描述与实际内容**

   PR 标题是"修复 PR #283 的测试问题"，但实际包含了 #283 的全部业务代码。Maintainer 第一眼看可能会以为只是测试改动，建议在描述里说明"这是基于 #283 的完整实现，附带 Gemini 指出的修复"，方便后续 review。

**关于合并路径建议：**

#437 和 #283 是 100% 重叠的，建议 maintainer 考虑直接合并 #437，同时关闭 #283 并在关闭说明里注明"由 #437 替代，感谢 @cw1427 的原始贡献"，保留对原作者的归属感谢。合并前如果能顺手处理一下上面的 #1（AutoConfiguration 启动副作用）和 #3（日志泄露）就更稳了。

不管最终怎么定，你这次的接力工作都很有价值，给 LDAP 功能补上了真实可用的实现 🙌

**Comment by jangrui** [comment]
**@XiaoSeS** 

感谢 Review，已按建议修复：

1. **AutoConfiguration 副作用** — 在 `application.yml` 中添加 `spring.ldap.urls: ""`，阻止 Spring Boot 默认 LDAP 自动配置在未启用时尝试初始化连接

2. **日志泄露密码风险** — 新增 `safeLogHost()` 方法，从 LDAP URL 中提取 host 时移除用户凭据部分，避免日志中打印密码

3. **资源泄露** — 在 `findUserDn()`、`authenticateLdap()`、`getUserAttributes()` 方法中添加 `try-finally` 块，确保 `DirContext` 和 `NamingEnumeration` 正确关闭

4. **连接超时** — 为所有 LDAP 连接添加 5 秒连接超时和 10 秒读取超时，防止无限等待

5. **LDAP 注入防护** — 新增 `isValidUsername()` 方法，对用户名进行正则校验，阻止恶意输入

6. **配置可维护性** — 新增 `LdapAutoConfiguration` 类，通过 `@ConditionalOnProperty` 实现零配置开关，仅设置 `skillhub.ldap.enabled=true` 即可启用

请再次 Review。

**Review by XiaoSeS — CHANGES_REQUESTED** [review]
> *This was generated by AI during triage.*

This is not safe to merge yet. The current CI and my Java 21 merge-tree build prove that the code compiles, but they do not cover the LDAP identity and provisioning behavior introduced by this PR.

1. **Blocking security issue: an LDAP login can take over an existing account by email collision.** In `LdapAuthService.findOrCreateLdapUser`, a successful LDAP bind followed by a matching `mail` value directly reuses any existing `UserAccount`. There is no LDAP identity binding or explicit account-link/merge check. An LDAP entry with the same email as a local/OAuth account therefore receives that account and its roles. LDAP users must be keyed by a stable provider subject such as `entryUUID`, `objectGUID`, or another configured immutable identifier, using the existing identity-binding/account-merge model. Email alone must not silently link accounts.

2. **The requested provisioning/sync contract is incomplete.** Issue #260 asks for configurable synchronized attributes. This implementation hard-codes `mail`, `displayName`, and `cn`, and does not update attributes on later logins. The no-mail fallback `ldap:{username}@internal` is also not a stable directory identity: usernames can change and the same username can exist in different directory branches.

3. **There is no functional LDAP or LDAPS verification.** The added tests mock `LdapAuthService`; `LdapAuthService` itself has no unit or integration coverage. Before merge, please verify at least OpenLDAP login, LDAPS trust/certificate handling, first-login provisioning, repeat login and attribute synchronization, duplicate-email behavior, invalid credentials, directory unavailable, and LDAP-disabled startup.

4. **Directory failures are hidden and user-facing message keys are missing.** Search, bind, attribute lookup, and attribute-read exceptions are broadly converted to `null`/`false`, so bad configuration or an unavailable LDAP server is reported like “user not found” or “invalid credentials”. Also, the `error.auth.ldap.*` keys used here do not exist in the message bundles. Please preserve safe diagnostic logging/exception classification and add localized messages.

5. `LdapTemplate` is configured and injected but never used; all operations duplicate connection setup through raw JNDI. Please either use the configured client consistently or remove the dead dependency/configuration and test the chosen lifecycle.

Validation performed against current `main` plus PR head `9e178711`:

- Java 21 full reactor compile/package: passed.
- LDAP-disabled Spring context via `LocalAuthControllerTest`: 13/13 passed.
- Existing `LocalAuthServiceTest` LDAP fallback cases: passed, but they only exercise a mocked LDAP service.

Because this is a large authentication feature rather than a simple test fix, I recommend updating the PR title/description and addressing the identity model plus real-directory tests before another review.

**Comment by XiaoSeS** [comment]
> *This was generated by AI during triage.*

维护状态说明：当前 HEAD `9e178711` 已在 2026-07-28 收到可操作的 Changes Requested。按新的历史 PR 处理策略，从该次 Review 起保留 14 天作者响应窗口（至 2026-08-11）。

在此期间维护者不会改写作者提交；如作者补交修改，将基于新 HEAD 继续 Review。若届时仍无回复，维护者可通过额外的、带 DCO sign-off 的提交接管修复，并按 `big-main` → 测试机验证的顺序处理，不会直接进入 `main`。

当前结论仍是阻塞，尚未进入 `big-main`，也未在测试机部署。

**Comment by XiaoSeS** [comment]
> *This was generated by AI during triage.*

补充一份可执行的改进方案，便于后续修复和重新验证：

1. **以稳定目录标识建立身份绑定。** 增加可配置的 LDAP subject 属性（OpenLDAP 可使用 `entryUUID`，AD 可使用 `objectGUID`），通过现有 `IdentityBinding` 以 `provider=ldap + subject` 查找账号；不要再把 email 当成身份主键。
2. **禁止邮箱静默合并。** LDAP identity 尚未绑定但 email 已属于本地/OAuth 账号时，应返回明确冲突并进入显式绑定/合并流程，不能直接继承该账号及其角色。
3. **修复无邮箱用户重复创建。** 首次登录保存 identity binding；后续登录必须先按 subject 命中同一账号，`ldap:{username}@internal` 只能作为占位邮箱，不能承担身份键职责。
4. **实现可配置的属性同步。** 至少支持 username、displayName、email 的属性映射，并明确首次创建与后续登录的同步策略；不能只在首次创建时硬编码读取 `mail/displayName/cn`。
5. **统一 LDAP 客户端和错误分类。** 使用已配置的 `LdapTemplate/ContextSource`，或删除未使用 Bean 后完整封装 JNDI；区分用户不存在、密码错误、目录不可用、TLS/证书错误，并补齐中英文消息键。
6. **补齐验证。** 单元测试覆盖邮箱碰撞、无邮箱重复登录、属性更新和禁用账号；集成测试覆盖 OpenLDAP、LDAPS、首次/重复登录、错误密码、目录不可用及 LDAP disabled 启动。

完成上述修改后，建议先合入 `big-main`，再在测试机使用临时 LDAP/LDAPS 容器跑完整链路；验证通过后才具备合并条件。当前 HEAD 不建议直接合并。

**Comment by FenjuFu** [comment]
Heads-up on a collision: **#672 (@XiaoSeS, opened 2026-07-31) also implements enterprise LDAP/AD login**, so this PR and that one overlap and would conflict if both landed.

- **This PR #437** (2026-05-14, based on #283): standalone `LdapAuthController` + `LdapBindingAppService`, LDAP as a fallback after local auth, explicit bind endpoint for email collisions.
- **#672**: LDAP/AD as an adapter inside the unified identity core (`DirectAuthService`), part of the federation design in #630.

Both do OpenLDAP `entryUUID` / AD `objectGUID` and both return 409 on email collision instead of silently merging — same problem, two designs.

A maintainer should pick one lane before either merges. @dongjiang1989 — standalone (#437) or folded into the federation core (#672)? The explicit-bind endpoint here (`POST /api/v1/auth/ldap/bind`) is a genuinely useful piece for the email-conflict flow and could be carried into whichever design wins.

Flagging so these converge rather than race — not voting on the direction.

**Comment by XiaoSeS** [comment]
维护者复核结论（保留 PR，不关闭）：

#437 的 LDAP 功能与已在 #672 采用的统一身份核心存在架构重叠。#437 直接在 `LocalAuthService` fallback 中按 LDAP 结果建号/绑定，#672 则把 LDAP/AD 放进统一的 provider adapter，由稳定 subject（OpenLDAP `entryUUID` / AD `objectGUID`）和 identity binding 负责首次登录、重复登录、显式 Link、Merge 与 profile policy。两条实现不应同时进入 `main`，否则会出现两套 LDAP 登录和账号绑定语义。

因此当前不建议把 #437 直接合入 `main` 或 `big-main`。#672 已在 `big-main` 完成香港隔离验证，覆盖 provider 目录、首次/重复登录、logout、错误密码/未知用户脱敏和真实 uid 改名后的稳定 userId；验证记录见 https://github.com/iflytek/skillhub/pull/672#issuecomment-5161586723。

#437 中仍有可移植的设计/测试价值，建议保留为参考：

- 显式 LDAP bind/link 入口，可作为统一 identity core 的 email collision / explicit link UX 设计输入；
- 目录不可用、TLS/证书和 LDAP disabled 启动的测试场景；
- 对多 OU 同名账号、属性同步和错误分类的边界覆盖。

如果继续维护本 PR，建议不要再增加第二套账号创建逻辑，而是把上述场景改写为统一 provider adapter / identity-link contract 的测试或设计补充。当前维持 Changes Requested，等待后续统一身份方案收敛。

**Comment by FenjuFu** [comment]
Thanks @XiaoSeS for the clear call — this resolves the collision I flagged. Agreed on the direction: LDAP folded into the unified identity core (#672 on `big-main`) rather than two parallel account-creation paths.

@jangrui — the useful parts of this PR aren't wasted. Per XiaoSeS's note, the explicit LDAP bind/link endpoint and the edge-case tests here (directory-unavailable, TLS/cert, LDAP-disabled startup, multi-OU same-name accounts, attribute sync, error classification) are exactly the kind of coverage the unified adapter will want. The most valuable next step would be to re-target those scenarios as tests / design input against the #672 identity-link contract, rather than maintaining a second `LocalAuthService`-fallback build path. Keeping this open as reference (Changes Requested) makes sense until the unified approach lands.

**Comment by jangrui** [comment]
感谢两位的反馈和复核。同意方向：LDAP 并入统一身份核心（#672/#630），不再维护第二条账号创建路径。

本 PR 的资产可供统一 adapter 参考：
- 显式 bind/link 端点 `POST /api/v1/auth/ldap/bind`
- 测试覆盖：目录不可用、LDAP disabled 启动、并发首登、稳定 subject 标识、多 OU 同名账号、TLS/证书、错误分类

后续如有推进，将以对 #672 identity-link contract 的测试/设计补充形式跟进。再次感谢 @cw1427 的原始贡献。
